components/homme/src/share/gllfvremap_test_mod.F90

```diff
@@ -0,0 +1,552 @@
+             else
+                call get_temperature(elem(ie), wr1, hvcoord, nt1)
+                global_shared_buf(ie,1) = sum(wr*(elem(ie)%derived%FT - wr1)**2)
+                global_shared_buf(ie,2) = sum(wr*wr1**2)                
+                wr1 = wr1*(elem(ie)%state%dp3d(:,:,:,nt1)/p0)**kappa
+                global_shared_buf(ie,4) = sum(wr(:,:,1)*elem(ie)%state%dp3d(:,:,1,nt1)*wr1(:,:,1))
+                wr1 = elem(ie)%derived%FT
+                wr1 = wr1*(elem(ie)%state%dp3d(:,:,:,nt1)/p0)**kappa
+                global_shared_buf(ie,3) = sum(wr(:,:,1)*elem(ie)%state%dp3d(:,:,1,nt1)*wr1(:,:,1))
+             end if
+          else
+             if (ftype == 0) then
+                elem(ie)%derived%FQ(:,:,:,q) = elem(ie)%state%Q(:,:,:,q) + &
+                     dt*elem(ie)%derived%FQ(:,:,:,q)/ &
+                     elem(ie)%state%dp3d(:,:,:,nt1)
+             end if
+             ! Extrema in level 1.
+             qmin2 = min(qmin2, minval(elem(ie)%derived%FQ(:,:,1,q)))
+             qmax2 = max(qmax2, maxval(elem(ie)%derived%FQ(:,:,1,q)))
+             ! Mass in level 1.
+             global_shared_buf(ie,3) = sum(wr(:,:,1)*elem(ie)%state%dp3d(:,:,1,nt1)* &
+                  elem(ie)%derived%FQ(:,:,1,q))
+             global_shared_buf(ie,4) = sum(wr(:,:,1)*elem(ie)%state%dp3d(:,:,1,nt1)* &
+                  two*elem(ie)%state%Q(:,:,1,q))
+             ! l2 error in volume.
+             global_shared_buf(ie,1) = &
+                  sum(wr*( &
+                  elem(ie)%derived%FQ(:,:,:,q) - &
+                  two*elem(ie)%state%Q(:,:,:,q))**2)
+             global_shared_buf(ie,2) = &
+                  sum(wr*elem(ie)%state%Q(:,:,:,q)**2)
+          end if
+       end do
+       call wrap_repro_sum(nvars=4, comm=hybrid%par%comm)
+       qmin1(q) = ParallelMin(qmin1(q), hybrid)
+       qmax1(q) = ParallelMax(qmax1(q), hybrid)
+       qmin2 = ParallelMin(qmin2, hybrid)
+       qmax2 = ParallelMax(qmax2, hybrid)
+       if (hybrid%masterthread) then
+          rd = sqrt(global_shared_sum(1)/global_shared_sum(2))
+          write(iulog, '(a,i3,a,i3,es12.4)') 'gfrt> test3 q l2', q, ' of', qsize, rd
+          b = max(abs(qmin1(q)), abs(qmax1(q)))
+          if (q <= qsize .and. qmin2 < qmin1(q) - 5*eps*b .or. &
+               qmax2 > qmax1(q) + 5*eps*b) then
+             write(iulog, '(a,i3,es12.4,es12.4,es12.4,es12.4,a)') 'gfrt> test3 q extrema', &
+                  q, qmin1(q), qmin2-qmin1(q), qmax2-qmax1(q), qmax1(q), ' ERROR'
+          end if
+          if (domass) then
+             a = global_shared_sum(3)
+             b = global_shared_sum(4)
+             if (abs(b - a) > 5*eps*abs(a)) then
+                write(iulog, '(a,i3,es12.4,es12.4,es12.4,a)') 'gfrt> test3 q mass', &
+                     q, a, b, abs(b - a)/abs(a), ' ERROR'
+             end if
+          end if
+       end if
+    end do
+  end subroutine run
+
+  subroutine gfr_check_api(hybrid, nets, nete, hvcoord, elem)
```

Comment:
What does this routine do?

Reply:
It runs a sequence of tests that are permitted to use gllfvremap_mod
only through its public API, i.e., the same routines called in the
full atmosphere. This is in contrast to the unit tests in
gllfvremap_mod.F90, which access internal functions. The API tests are
in three parts, as follows (copy-pasted from comments):

```
    !! Test 1.                                                                                                                                                                                                                                                   
    ! Test physgrid API.                                                                                                                                                                                                                                         
    !   Test that if tendency is 0, then the original field is                                                                                                                                                                                                   
    ! recovered with error eps ~ machine precision.                                                                                                                                                                                                              
    !   OOA for pgN should min(N, 2). The OOA is limited to 2 because                                                                                                                                                                                            
    ! of the way the tendencies are created. Instead of setting the FV                                                                                                                                                                                           
    ! value to the average over the subcell, the sampled value at the                                                                                                                                                                                            
    ! cell midpoint (ref midpoint mapped to sphere) is used. This                                                                                                                                                                                                
    ! creates a 2nd-order error. Tests 2 and 3 test whether the remaps                                                                                                                                                                                           
    ! achieve design OOA, in particular OOA 3 for pg3 u, v, T, phis                                                                                                                                                                                              
    ! fields.                                                                                                                                                                                                                                                    

    !! Test 2.                                                                                                                                                                                                                                                   
    ! Test topo routines. For pgN, phis should have OOA min(N,2)                                                                                                                                                                                                 
    ! because the field is limited.                                                                                                                                                                                                                              

    !! Test 3.                                                                                                                                                                                                                                                   
    ! Test FV fields that were not covered in the previous tests. This                                                                                                                                                                                           
    ! is done by copying them to look like tendencies.                                                                                                                                                                                                           
    !   For pg4, u,v,T should have l2 errors that are eps because pg4                                                                                                                                                                                            
    ! reconstructions the fields exactly, even with DSS.                                                                                                                                                                                                         
    !   For pg4, q should have l2 errors that converge at OOA >= 2. 2                                                                                                                                                                                            
    ! is the formal OOA b/c of the limiter. The limiter acts on the                                                                                                                                                                                              
    ! tendency, which in this case is exactly what is being examined.                                                                                                                                                                                            
    !   For pgN, N=1,2,3, u,v,T should have OOA N.                                                                                                                                                                                                               
    !   For pgN, q should have OOA min(N,2).                                                                                                                                                                                                                     
```